File: androidApp/src/main/java/com/badmintontracker/android/localvideo/LocalVideoSection.kt
package com.badmintontracker.android.localvideo

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import coil3.compose.AsyncImage
import com.badmintontracker.android.cliplist.DrawerList
import com.badmintontracker.android.cliplist.DrawerRowContainer
import com.badmintontracker.android.cliplist.formatDate
import com.badmintontracker.android.ui.components.ShuttlButton
import com.badmintontracker.android.ui.components.ShuttlButtonVariant
import com.badmintontracker.android.ui.theme.ShuttlRadius
import com.badmintontracker.android.ui.theme.ShuttlTheme
import com.badmintontracker.shared.localvideo.LocalVideoEntry
import kotlinx.datetime.Instant

/**
 * "On this phone" section rendered inside the matches LazyColumn.
 *
 * The mock draws these as filled cards rather than list rows: this is the
 * shortest path from "I just filmed a match" to "analyze it", so it gets the
 * one raised surface in the drawer while the match rows below stay quiet.
 *
 * [onRemoveRequest] asks the host screen to confirm; actual removal happens
 * there, so swipes must not dismiss the row.
 */
fun LazyListScope.localVideoSection(
    rows: List<LocalVideoRow>,
    header: @Composable (String) -> Unit,
    onRowClick: (LocalVideoEntry) -> Unit,
    onAnalyzeClick: (LocalVideoRow) -> Unit,
    onRemoveRequest: (LocalVideoEntry) -> Unit,
    onEditDetails: (LocalVideoEntry) -> Unit,
    /**
     * Null when this video has no stored player track.
     *
     * The heatmap used to be reachable only from the banner of the run that
     * produced it, so it vanished with the process even though the track was on
     * disk - half an hour of analysis with no way back to it.
     */
    onOpenHeatmap: ((LocalVideoEntry) -> Unit)? = null,
    hasHeatmap: (LocalVideoEntry) -> Boolean = { false },
    onOpenLocalClips: ((LocalVideoEntry) -> Unit)? = null,
    localClipCount: (LocalVideoEntry) -> Int = { 0 },
) {
    if (rows.isEmpty()) return
    item(key = "header-local") { header("On this phone") }
    items(rows, key = { "local-${it.entry.id}" }) { row ->
        DrawerRowContainer(
            shape = RoundedCornerShape(ShuttlRadius.large),
            fill = ShuttlTheme.extended.bgTertiary,
            // Mid-pipeline rows pass null: removing would delete the file under
            // the active upload, or out from under the device run still
            // decoding it, and swallow the run's outcome.
            swipeLabel = "Remove".takeIf { row.canRemove },
            onSwiped = { onRemoveRequest(row.entry); false },
        ) {
            LocalVideoRowItem(
                row = row,
                onClick = { onRowClick(row.entry) },
                onAnalyze = { onAnalyzeClick(row) },
                onRemove = { onRemoveRequest(row.entry) },
                onEditDetails = { onEditDetails(row.entry) },
                onOpenHeatmap = onOpenHeatmap?.takeIf { hasHeatmap(row.entry) }
                    ?.let { open -> { open(row.entry) } },
                localClips = localClipCount(row.entry).takeIf { it > 0 },
                onOpenLocalClips = onOpenLocalClips?.let { open -> { open(row.entry) } },
            )
        }
    }
}

@Composable
private fun LocalVideoRowItem(
    row: LocalVideoRow,
    onClick: () -> Unit,
    onAnalyze: () -> Unit,
    onRemove: () -> Unit,
    onEditDetails: () -> Unit,
    onOpenHeatmap: (() -> Unit)? = null,
    localClips: Int? = null,
    onOpenLocalClips: (() -> Unit)? = null,
) {
    val entry = row.entry
    var menuOpen by remember { mutableStateOf(false) }
    // The menu renders when either action applies; each item is gated on its
    // own rule, so a mid-pipeline row that can do neither shows no menu at all.
    val hasMenu = row.canRemove || row.canEditDetails

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(
                start = DrawerList.rowPaddingH,
                // An IconButton carries 12dp of its own padding inside a 48dp
                // box. Backing that out here is what puts the glyph the same
                // 16dp from the card's edge as the thumbnail on the other side,
                // instead of an optical 28dp that reads as a misaligned card.
                end = if (hasMenu) DrawerList.rowPaddingH - 12.dp else DrawerList.rowPaddingH,
                top = DrawerList.rowPaddingV,
                bottom = DrawerList.rowPaddingV,
            ),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            AsyncImage(
                model = entry.uri,
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier
                    .size(DrawerList.thumbWidth, DrawerList.thumbHeight)
                    .clip(RoundedCornerShape(ShuttlRadius.small))
                    // A thumbnail that has not decoded yet would otherwise be a
                    // card-coloured hole; this reads as an empty slot instead.
                    .background(MaterialTheme.colorScheme.outlineVariant),
            )
            Spacer(Modifier.width(DrawerList.gap))
            Column(Modifier.weight(1f)) {
                Text(
                    row.primaryText,
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onBackground,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    "${row.durationText} · " +
                        formatDate(Instant.fromEpochMilliseconds(entry.addedAtEpochMs)),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                row.statusText?.let { status ->
                    Text(
                        text = status,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        // Two, not one: the drawer leaves this column about 118dp
                        // once the thumbnail and the menu have taken theirs. One
                        // line is enough for every label the row shows at the
                        // default font scale; at a larger one this wraps instead
                        // of truncating mid-word.
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
            if (row.isBusy && !row.canAnalyze) {
                // Either pipeline, not just the cloud one: a device run leaves the
                // stage alone, so asking isAnalysisRunning(stage) here left the row
                // showing nothing at all while the phone was analysing it.
                //
                // Settled stages (e.g. ANALYZED) show neither ring nor button -
                // the status text already says what happened.
                // Indeterminate even for a run whose fraction is known: at 16dp a
                // determinate ring spends the first minutes looking like an empty
                // grey circle, which reads as "stalled" rather than "2% done". The
                // motion is what says the phone is working; the number lives on
                // Home's own analysis banner and in the chrome indicator, both of
                // which have room to draw it.
                Spacer(Modifier.width(DrawerList.gap))
                CircularProgressIndicator(
                    modifier = Modifier.size(16.dp),
                    strokeWidth = 2.dp,
                )
            }
            if (hasMenu) {
                // The menu must share a Box with its anchor: DropdownMenu positions
                // itself relative to its parent, not the IconButton.
                Box {
                    IconButton(onClick = { menuOpen = true }) {
                        Icon(Icons.Default.MoreVert, contentDescription = "Local video menu")
                    }
                    DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                        if (localClips != null && onOpenLocalClips != null) {
                            DropdownMenuItem(
                                text = { Text("Clips on this phone ($localClips)") },
                                onClick = { menuOpen = false; onOpenLocalClips() },
                            )
                        }
                        if (onOpenHeatmap != null) {
                            DropdownMenuItem(
                                text = { Text("Player heatmap") },
                                onClick = { menuOpen = false; onOpenHeatmap() },
                            )
                        }
                        if (row.canEditDetails) {
                            DropdownMenuItem(
                                text = { Text("Edit details") },
                                onClick = { menuOpen = false; onEditDetails() },
                            )
                        }
                        if (row.canRemove) {
                            DropdownMenuItem(
                                text = { Text("Remove from app") },
                                onClick = { menuOpen = false; onRemove() },
                            )
                        }
                    }
                }
            }
        }
        if (row.canAnalyze) {
            // Its own line, not the trailing slot the mock draws it in. The
            // mock's card carries a thumbnail, two lines and one pill; this one
            // also carries the overflow menu, and squeezing a pill in beside it
            // left the title about 38dp wide - roughly four characters. Full
            // width instead, which is also how Home states its two primary
            // actions, so the card reads as one clear next step.
            Spacer(Modifier.height(DrawerList.gap))
            ShuttlButton(
                text = row.analyzeLabel,
                onClick = onAnalyze,
                variant = ShuttlButtonVariant.Primary,
                compact = true,
                modifier = Modifier
                    .fillMaxWidth()
                    // Cancels the menu compensation above, so the pill ends flush
                    // with the thumbnail's left edge and the card's right inset.
                    .padding(end = if (hasMenu) 12.dp else 0.dp),
            )
        }
    }
}

/** Modal shown when a local video's analysis ends in a failure or a no-rallies result. */
@Composable
fun AnalyzeResultDialog(
    entry: LocalVideoEntry,
    onRetry: () -> Unit,
    onDismiss: () -> Unit,
) {
    val message = entry.failureMessage ?: "Unknown error"
    val noRallies = message.contains("no rallies", ignoreCase = true)
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(if (noRallies) "No rallies found" else "Analysis failed") },
        text = { Text(message) },
        confirmButton = { TextButton(onClick = onRetry) { Text("Retry") } },
        dismissButton = { TextButton(onClick = onDismiss) { Text("Close") } },
    )
}

File: iosApp/Sources/LocalVideo/LocalVideoSection.swift
import Shared
import SwiftUI

/// Navigation value for opening the local player (distinct from the String
/// destination used for remote match ids).
struct LocalPlayerRoute: Hashable {
    let entryId: String
}

/// The "On this phone" card.
///
/// The mock draws these as filled cards rather than list rows: this is the
/// shortest path from "I just filmed a match" to "analyze it", so it gets the
/// one raised surface in the drawer while the match rows below stay quiet.
/// Mirrors androidApp's LocalVideoRowItem.
struct LocalVideoRowView: View {
    let entry: LocalVideoEntry
    let thumbnails: LocalThumbnails
    let progress: AnalyzeProgress?
    /// What the on-device pipeline is doing with this video. Asked alongside the
    /// cloud stage rather than instead of it - see `LocalVideoStatus.canAnalyze`.
    var device: LocalAnalysisState = .idle
    /// Pushes `LocalPlayerRoute` on the host's behalf. A plain `Button` rather
    /// than `NavigationLink(value:)`: the host owns the destination as an
    /// `item:` binding now (Home owns every destination this row's list used
    /// to carry), so the route is reported upward instead of resolved through
    /// the navigation environment.
    let onTap: () -> Void
    let onAnalyze: () -> Void
    let onRemove: () -> Void
    let onEditDetails: () -> Void
    /// Opens the rallies an on-device run cut, or nil when it produced none.
    ///
    /// The clips used to be reachable only from the state of the run that made
    /// them, so they went with the process even though tens of megabytes of
    /// re-encoded video were on disk - half an hour of analysis with no way back
    /// to what it produced. Mirrors androidApp's `onOpenLocalClips`.
    var localClips: Int? = nil
    var onOpenLocalClips: (() -> Void)? = nil
    /// Opens the heatmap of a stored track, on the same footing and for the same
    /// reason. Nil when this entry has no track.
    var onOpenHeatmap: (() -> Void)? = nil

    private var subtitle: String { localVideoSubtitle(entry) }
    private var canAnalyze: Bool { LocalVideoStatus.canAnalyze(stage: entry.stage, device: device) }
    private var canRemove: Bool { LocalVideoStatus.canRemove(stage: entry.stage, device: device) }
    /// The menu renders when ANY of its four items applies; each is gated on its
    /// own rule, so a mid-pipeline row that can do none shows no menu at all.
    /// The two analysis items count: gating on remove and edit alone would hide
    /// the clips and the heatmap on exactly the rows most likely to have them.
    private var hasMenu: Bool {
        canRemove || LocalVideoStatus.canEditDetails(stage: entry.stage)
            || onOpenHeatmap != nil || (localClips != nil && onOpenLocalClips != nil)
    }
    /// A ring turns for work that is moving, on either pipeline.
    ///
    /// A paused device run is the exception, and it is the case this row is most
    /// likely to be SEEN in: the app switcher snapshots the scene on the way
    /// out, so the card a coach checks while he is away is this one. A ring
    /// turning there over a run getting no CPU would contradict the line beside
    /// it, which reads "Paused - keep Shuttl open".
    private var spins: Bool {
        if case .paused = device { return false }
        return (LocalVideoStatus.isRunning(stage: entry.stage) || isDeviceRunInFlight(device))
            && !canAnalyze
    }

    var body: some View {
        Button(action: onTap) {
            VStack(alignment: .leading, spacing: DrawerList.gap) {
                HStack(spacing: DrawerList.gap) {
                    Group {
                        if let image = thumbnails.images[entry.id] {
                            Image(uiImage: image).resizable().aspectRatio(contentMode: .fill)
                        } else {
                            // A thumbnail that has not decoded yet would
                            // otherwise be a card-coloured hole; this reads as
                            // an empty slot instead.
                            Shuttl.border
                        }
                    }
                    .frame(width: DrawerList.thumbWidth, height: DrawerList.thumbHeight)
                    .clipShape(RoundedRectangle(cornerRadius: ShuttlRadius.small))
                    .task { await thumbnails.load(for: entry) }

                    VStack(alignment: .leading, spacing: 3) {
                        Text(entry.title ?? entry.displayName)
                            .shuttlType(ShuttlType.titleMedium)
                            .foregroundStyle(Shuttl.text)
                            .lineLimit(1)
                        Text(subtitle)
                            .shuttlType(ShuttlType.bodySmall)
                            .foregroundStyle(Shuttl.textSecondary)
                            .lineLimit(1)
                        if let status = LocalVideoStatus.rowStatus(
                            stage: entry.stage, progress: progress, device: device
                        ) {
                            Text(status)
                                .shuttlType(ShuttlType.bodySmall)
                                .foregroundStyle(Shuttl.textSecondary)
                                // Two, not one: the drawer leaves this column
                                // about 118pt once the thumbnail and the menu
                                // have taken theirs.
                                .lineLimit(2)
                        }
                    }
                    Spacer(minLength: 0)
                    if spins {
                        // Settled stages (e.g. ANALYZED) show neither ring nor
                        // button - the status text already says what happened.
                        ProgressView().controlSize(.small)
                    }
                    if hasMenu {
                        Menu {
                            if let localClips, let onOpenLocalClips {
                                Button("Clips on this phone (\(localClips))") { onOpenLocalClips() }
                            }
                            if let onOpenHeatmap {
                                Button("Player heatmap") { onOpenHeatmap() }
                            }
                            if LocalVideoStatus.canEditDetails(stage: entry.stage) {
                                Button("Edit details") { onEditDetails() }
                            }
                            if canRemove {
                                Button("Remove from app", role: .destructive) { onRemove() }
                            }
                        } label: {
                            Image(systemName: "ellipsis")
                                .foregroundStyle(Shuttl.textSecondary)
                                // The same 44x44 minimum every other trailing
                                // control in this list carries, so the glyph is
                                // tappable without widening the card's inset.
                                .frame(width: 44, height: 44, alignment: .trailing)
                        }
                        .buttonStyle(.borderless)
                        .accessibilityLabel("Local video menu")
                    }
                }
                if canAnalyze {
                    // Its own line, not the trailing slot the mock draws it in.
                    // The mock's card carries a thumbnail, two lines and one
                    // pill; this one also carries the overflow menu, and
                    // squeezing a pill in beside it left the title about four
                    // characters wide. Full width instead, which is also how
                    // Home states its two primary actions, so the card reads as
                    // one clear next step.
                    Button(LocalVideoStatus.analyzeButtonLabel(stage: entry.stage)) { onAnalyze() }
                        .shuttlType(ShuttlType.labelMedium)
                        .foregroundStyle(Shuttl.onAccent)
                        .lineLimit(1)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 10)
                        .background(Shuttl.accent)
                        .clipShape(Capsule())
                        .buttonStyle(.borderless)
                }
            }
            .padding(.horizontal, DrawerList.rowPaddingH)
            .padding(.vertical, DrawerList.rowPaddingV)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Shuttl.bgTertiary)
            .clipShape(RoundedRectangle(cornerRadius: ShuttlRadius.large))
            // Without this, the Button's hit area is only its children's -
            // the Spacer in the middle has none of its own - so the empty
            // stretch between the text and the trailing control would go dead
            // and silently stop opening the video. NavigationLink gave the
            // whole row a hit area for free; a Button does not.
            .contentShape(RoundedRectangle(cornerRadius: ShuttlRadius.large))
        }
        .buttonStyle(.plain)
    }
}

extension LocalVideoEntry: Identifiable {}
